packages/apollo-server-core/src/caching.ts

```diff
@@ -1,4 +1,44 @@
+import { ExecutionResult } from 'graphql';
+
 export interface PersistedQueryCache {
   set(key: string, data: string): Promise<any>;
   get(key: string): Promise<string | null>;
 }
+
+export function calcualteCacheControlHeaders(
+  responses: Array<ExecutionResult & { extensions?: Record<string, any> }>,
+) {
+  let maxAge = Number.MAX_VALUE;
+  let publicOrPrivate = 'public';
+
+  for (let i = 0; i < responses.length; i++) {
```

Comment:
So the key thing in the engineproxy calculation of cacheControl is you need to put a hint on every top-level data piece in order for it to count. See the rootDataKey loop in parseCacheControl.